optimism/src/keccak/mod.rs

```diff
@@ -70,6 +79,77 @@ pub enum Constraint {
     IotaStateG(usize),
 }
 
+#[allow(dead_code)]
+/// The Keccak circuit
+pub(crate) type KeccakCircuit<F> = Circuit<ZKVM_KECCAK_COLS, Steps, F>;
+
+#[allow(dead_code)]
+impl<F: Field> KeccakCircuit<F> {
+    /// Create a new Keccak circuit
+    pub fn new(domain_size: usize) -> Self {
+        let mut circuit = Self {
+            witness: HashMap::new(),
+            constraints: Default::default(),
+            lookups: Default::default(),
+        };
+
+        for step in [
+            Round(0),
+            Sponge(Absorb(First)),
+            Sponge(Absorb(Middle)),
+            Sponge(Absorb(Last)),
+            Sponge(Absorb(Only)),
+            Sponge(Squeeze),
+        ] {
+            circuit.witness.insert(
+                step,
+                Witness {
+                    cols: Box::new(std::array::from_fn(|_| Vec::with_capacity(domain_size))),
+                },
+            );
+            circuit
+                .constraints
+                .insert(step, KeccakEnv::constraints_of(step));
+            circuit.lookups.insert(step, KeccakEnv::lookups_of(step));
+        }
+        circuit
+    }
+
+    /// Add a witness row to the circuit
+    pub(crate) fn push_row(&mut self, step: Steps, row: &[F; ZKVM_KECCAK_COLS]) {
```

Comment:
This just pushes a witness row to the full instance of the circuit being stored.